src/compiler/parser.ts

```diff
@@ -4040,7 +4058,13 @@ namespace ts {
                 else if (token() === SyntaxKind.EndOfFileToken) {
                     // If we hit EOF, issue the error at the tag that lacks the closing element
                     // rather than at the end of the file (which is useless)
-                    parseErrorAtPosition(openingTagName.pos, openingTagName.end - openingTagName.pos, Diagnostics.JSX_element_0_has_no_corresponding_closing_tag, getTextOfNodeFromSourceText(sourceText, openingTagName));
+                    if (isJsxOpeningElement(openingTag)) {

```

Comment:
Reading this whole thing it seems like it'd be clearer if we wrote e.g. `if (isJsxFragment(openingTag))` instead of `isJsxOpeningElement` since the former is more the thing we're special-casing

Reply:
Restructured.